openduck-py/openduck_py/routers/voice.py

```diff
@@ -129,94 +129,96 @@ def interrupt(self, task: asyncio.Task):
+                )
+            ).scalar_one_or_none()
+            if chat is None:
+                chat = DBChatHistory(
+                    session_id=self.session_id, history_json={"messages": [system_prompt]}
+                )
+                db.add(chat)
+            messages = chat.history_json["messages"]
+            messages.append(new_message)
+            response = await generate({"messages": messages}, [], "gpt-35-turbo-deployment")
+            response_message = response.choices[0].message
+            completion = response_message.content
+            await log_event(db, self.session_id, "generated_completion", meta={"text": completion})
+
+            t_gpt = time()
+
+            print(f"Used {response.usage.prompt_tokens} prompt tokens and {response.usage.completion_tokens} completion tokens")
+
+            if "$ECHO" in completion:
+                print("Echo detected, not sending response.")
+                return
 
-        messages.append(
-            {"role": response_message.role, "content": completion}
-        )
-        chat.history_json["messages"] = messages
-        await db.commit()
+            messages.append(
+                {"role": response_message.role, "content": completion}
+            )
+            chat.history_json["messages"] = messages
+            await db.commit()
 
-        normalized = normalize_text(response_message.content)
-        await log_event(db, self.session_id, "normalized_text", meta={"text": normalized})
-        t_normalize = time()
-        sentences = re.split(r"(?<=[.!?]) +", normalized)
-        for sentence in sentences:
-            audio_chunk_bytes = await loop.run_in_executor(None, _inference, sentence)
-            await log_event(db, self.session_id, "generated_tts", audio=np.frombuffer(audio_chunk_bytes, dtype=np.int16))
-            await websocket.send_bytes(audio_chunk_bytes)
+            normalized = normalize_text(response_message.content)
+            await log_event(db, self.session_id, "normalized_text", meta={"text": normalized})
+            t_normalize = time()
+            sentences = re.split(r"(?<=[.!?]) +", normalized)
+            for sentence in sentences:
+                audio_chunk_bytes = await loop.run_in_executor(None, _inference, sentence)
+                await log_event(db, self.session_id, "generated_tts", audio=np.frombuffer(audio_chunk_bytes, dtype=np.int16))
+                await websocket.send_bytes(audio_chunk_bytes)
 
-        t_styletts = time()
+            t_styletts = time()
 
-        print("Whisper", t_whisper - t0)
-        print("GPT", t_gpt - t_whisper)
-        print("Normalizer", t_normalize - t_gpt)
-        print("StyleTTS2 + sending bytes", t_styletts - t_normalize)
+            print("Whisper", t_whisper - t0)
+            print("GPT", t_gpt - t_whisper)
+            print("Normalizer", t_normalize - t_gpt)
+            print("StyleTTS2 + sending bytes", t_styletts - t_normalize)
+        await db.close()
```

Comment:
This line is not necessary - `db.close()` is called implicitly by the context manager.